### PR TITLE
AreasWithHighwayTagsCheck enhancements

### PR DESCRIPTION
**Pull request by nachtm**
### Description:

This is a series of enhancements to AreasWithHighwayTagsCheck. 
 - Previously, only areas were being checked. This missed out on a large number of OSM ways that were tagged with area=yes and some invalid highway value. 
 - In addition, there are some highway tags that are valid in conjunction with area=yes that were previously being flagged. Those are no longer flagged.
 - Finally, instead of just flagging the particular edge in question, we gather all connected edges with improper combinations of area and highway tags. This makes the output of this check easier to understand for people who are familiar with OSM but not Atlas. This may also reduce the flag count in some cases, since connected improperly tagged edges will be flagged as a single unit now.

In short, the new logic is as follows: for any area or edge, if that object has an area=yes tag and a highway tag that is not one of `{services, service, rest area, pedestrian, platform}`, flag it and all connected edges that share a similar problem.

### Potential Impact:

Changed flag counts for AreasWithHighwayTags.

### Unit Test Approach:

Added unit tests to cover edges and a couple of new area cases.

### Test Results:

All tests passing.

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java [hunk 1]
```diff
@@ -34,34 +45,77 @@ public AreasWithHighwayTagCheck(final Configuration configuration)
...
-            return Optional.of(this.createFlag(object,
-                    this.getLocalizedInstruction(localizedInstruction, object.getOsmIdentifier())));
-
+            return object.getTag(HighwayTag.KEY)
+                    .map(tagString -> HighwayTag.valueOf(tagString.toUpperCase()))
+                    // If the tag isn't one of the validHighwayTags, we want to flag it.
+                    .filter(tag -> isUnacceptableAreaHighwayTagCombination(object, tag)).map(tag ->
+                    {
+                        final String instruction;
+                        if (isNotOsmStandard(tag))
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java]: Why are these flagged differently?
Also, nit: this.isNotOsmStandard

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java [hunk 2]
```diff
@@ -34,34 +45,77 @@ public AreasWithHighwayTagCheck(final Configuration configuration)
...
     @Override
     protected Optional<CheckFlag> flag(final AtlasObject object)
     {
-        // See configuration for initial check filter; highway=pedestrian & area=yes is a valid
-        // Area/Highway tag combination.
-        final Area area = (Area) object;
-        final Optional<String> highwayTag = area.getTag(HighwayTag.KEY);
-        // First check if highway tag exists
-        if (highwayTag.isPresent())
+        if (!this.isFlagged(object.getOsmIdentifier()))
```
Comment (Bentleysb) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java]: We can handle this in validCheckForObject.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java [hunk 3]
```diff
@@ -22,9 +29,13 @@
 public class AreasWithHighwayTagCheck extends BaseCheck<Long>
 {
     private static final List<String> FALLBACK_INSTRUCTIONS = Arrays.asList(
-            "Area with OSM ID {0,number,#} contains invalid highway tag.",
-            "Area with OSM ID {0,number,#} is missing area tag.");
+            "The way with OSM ID {0,number,#} has a highway value which does not match the OSM standard.",
+            "The way with OSM ID {0,number,#} has a highway value of {1}, which should not have an area=yes tag.");
     private static final long serialVersionUID = 3638306611072651348L;
+    private static final EnumSet<HighwayTag> validHighwayTags = EnumSet.of(HighwayTag.SERVICES,
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java]: nit:`VALID_HIGHWAY_TAGS`

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java [hunk 4]
```diff
@@ -22,9 +29,13 @@
...
+            HighwayTag.SERVICE, HighwayTag.REST_AREA, HighwayTag.PEDESTRIAN, HighwayTag.PLATFORM);
+    private static final EnumSet<HighwayTag> notToStandardHighwayTags = EnumSet
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java]: nit : Is the `To` in the variable name a typo? Also constant could be in snake case.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheckWalker.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Simple walker to grab all connected edges which have a bad combination of area=yes and some
+ * highway tag.
+ *
+ * @author nachtm
+ */
+public class AreasWithHighwayTagCheckWalker extends EdgeWalker
+{
+
+    private static final Predicate<Edge> isBadEdgeWithAreaTag = edge -> AreasWithHighwayTagCheck
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheckWalker.java]: nit: `IS_BAD_EDGE_WITH_AREA_TAG`

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheckWalker.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            .isUnacceptableAreaHighwayTagCombination(edge, edge.highwayTag());
+
+    private static final Function<Edge, Stream<Edge>> getNeighborStream = edge -> edge
```
Comment (sayas01) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheckWalker.java]: nit: snake case

**Review by sayas01 — COMMENTED** [review]
Nice job @nachtm ! I just have a few comments regarding formatting of constants.

**Review by Bentleysb — CHANGES_REQUESTED** [review]
This looks pretty good, just a couple of little things and a question.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java (hunk 1 above)
Comment (nachtm) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java]: @kkonishi2 or @MonicaBrandeis can probably weigh in with a fuller explanation. The way I understand it, the non-standard tags are things that might seem reasonable, but should be tagged differently (highway=footway should actually be highway=pedestrian, for example).

**Review by sayas01 — APPROVED** [review]
LGTM!

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java (hunk 1 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java]: If that is the case it might be worth clarifying this in the instruction. Perhaps something where we give a suggested change like: "The way with OSM ID 1234567 has a highway value of highway=footway, which should not have an area=yes tag. Consider changing the highway tag to highway=pedestrian"

Comment (kkonishi2) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/AreasWithHighwayTagCheck.java]: I think Bentley's suggestion is fine. The more I think about it, the more I feel it's probably alright in this instance to suggest pedestrian. We run the risk of suggesting an incorrect fix when there should be no highway tag, however, with this check I think the majority of footway tags should probably just be pedestrian anyway.

**Review by Bentleysb — APPROVED** [review]
LGTM

**Review by danielduhh — APPROVED** [review]
Looks good - nice reviews everyone
